jetty9-compat-base/src/main/java/com/google/apphosting/vmruntime/jetty9/AppengineApiConfiguration.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.apphosting.vmruntime.jetty9;
+
+import org.eclipse.jetty.util.URIUtil;
+import org.eclipse.jetty.util.resource.Resource;
+import org.eclipse.jetty.webapp.AbstractConfiguration;
+import org.eclipse.jetty.webapp.WebAppClassLoader;
+import org.eclipse.jetty.webapp.WebAppContext;
+
+import java.util.logging.Level;
+
+public class AppengineApiConfiguration extends AbstractConfiguration {
+
+  // A class to check if the GAE API is available.
+  public static final String GAE_CHECK_CLASS = "com.google.appengine.api.ThreadManager";
+
+  // Hide the all container classes from the webapplication
+  // TODO update to use '.' when supported by Jetty
+  private static final String[] SERVER_CLASSES = {
+    "com.", 
+    "javax.", 
+    "org.", 
+    "mozilla."
+  };
+
+  private static final String[] SHARED_CLASSES = {
+    // Expose the GAE API classes
+    "com.google.appengine.api.LifecycleManager",
+    "com.google.apphosting.api.ApiProxy",
+    "com.google.apphosting.api.ApiStats",
+    "com.google.apphosting.api.CloudTrace",
+    "com.google.apphosting.api.CloudTraceContext",
+    "com.google.apphosting.api.DeadlineExceededException",
+    "com.google.apphosting.runtime.SessionData",
+    "com.google.apphosting.runtime.UncatchableError",
+
+    // Expose the standard APIs that are provided by the container
+    "javax.servlet.",
```

Comment:
Currently I'm handling all the shared classes the same: they are excluded from the serverClasses (so they are not hidden from the webapp) and they are included in the systemClasses (so they are protected from being overridden by the webapp).

In terms of the different reasons packages are shared, I have just done that by grouping in the source.

Now there are certainly packages in the shared list that need not be protected from replacement (ie they don't need to be system classes), so for example `javax.xml.` could probably be safely replaced by the webapp.   But then there are definitely others that must be protected, like `javax.el` which is shared between the webapp and the JSP implementation.

So without a clear policy or mechanism to determine should a package be protected or not, I've just applied a single shared white list.

I like white lists as they tend to be less vulnerable to accidental change as dependencies change, but in this particular case it may be simpler to revert back to a black list - ie to explicitly hide only those packages (eg `javax.cache.`) that the container requires but are not JVM provided classes.

The resulting list will definitely be shorter, but will be subject to change as we modify the compat implementation and we would be best to generate it rather than author it.

I think the white list is much better for having an explicitly clear policy on all packages.  Black list will be more flexible but less explicit and require more maintenance.         So I've got no clear preference, but Black list is the more normal container solution.